In [1]:
import pandas as pd
import numpy as np
import darkchem
import math

Using TensorFlow backend.


#### Several reactions of Reduction: `Benzene` `1-Butene` `2-Butene` `1-Hexene`

In [2]:
# Load reactants 
smi_in = {'reactants':['c1ccccc1','CCC=C','C/C=C/C','CCCCC=C']}
# Load products  
smi_out = {'products':['C1CCCCC1','CCCC','CCCC','CCCCCC']}       
# Transform to DataFrame
smi_in = pd.DataFrame(smi_in)  
smi_out = pd.DataFrame(smi_out)
# Represent reactants in the form of vector
rct = [darkchem.utils.struct2vec(smi) for smi in smi_in['reactants']]  
# Represent prodcuts in the form of vector
prod = [darkchem.utils.struct2vec(smi) for smi in smi_out['products']] 

In [3]:
# Put those vectors into DataFrame
rct = pd.DataFrame(rct)
prod = pd.DataFrame(prod)

In [4]:
# Calcualte the lengh of each reaction path
s = 0
d = []
for i in range(4):
    for j in range(100):
        s += (rct.loc[i][j] - prod.loc[i][j])**2
    ds = np.sqrt(s)
    d.append(ds)
print('The lengh of each reaction path =', d)

The lengh of each reaction path = [4.898979485566356, 21.18962010041709, 37.94733192202055, 43.18564576337837]


In [5]:
# Calculate the dot product of each reaction
s2 = 0
d2= []
for i in range(4):
    for j in range(100):
        s2 += (rct.loc[i][j] * prod.loc[i][j])
    #ds = np.sqrt(s)
    d2.append(s2)


# Calculate the product of each reactant and product
s3 = 0
s4 = 0
d3= []
for i in range(4):
    for j in range(100):
        s3 += (rct.loc[i][j])**2
        s4 += (prod.loc[i][j])**2
    dd = np.sqrt(s3)
    dx = np.sqrt(s4)
    xx = dd*dx
    d3.append(xx)

# Insert the formula of Cosine theorem
a = []
for i in range(4):
    cos = d2[i]/d3[i]  #this is the value of cos(sita)
    angle = math.acos(cos)  #this is the value of sita
    a.append(angle)
print('The angle between the vectors of reactants and products of each reactions =')
print(a)

The angle between the vectors of reactants and products of each reactions =
[0.008055875995473142, 0.4627571983462675, 0.5909134494895875, 0.5579381975667936]


#### Reduction
| Index |Benzene |1-Butene|2-Butene|1-Hexene|
|-------|--------|--------|--------|--------|
| Length| 4.89897|21.18962|37.94733|43.18564|
| Angle |0.008055|0.462757|0.590913|0.557938|